src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,446 @@
+        }
+
+        private static AttributeData[] FindAllPlatformAttributesApplied(ImmutableArray<AttributeData> immediateAttributes, INamedTypeSymbol parent)
+        {
+            using var builder = ArrayBuilder<AttributeData>.GetInstance();
+            foreach (AttributeData attribute in immediateAttributes)
+            {
+                if (s_osPlatformAttributes.Contains(attribute.AttributeClass.Name))
+                {
+                    builder.Add(attribute);
+                }
+            }
+
+            while (parent != null)
+            {
+                var current = parent.GetAttributes();
+                foreach (var attribute in current)
+                {
+                    if (s_osPlatformAttributes.Contains(attribute.AttributeClass.Name) &&
+                        TargetPlatformAttribute != attribute.AttributeClass.Name)
+                    {
+                        builder.Add(attribute);
+                    }
+                }
+                parent = parent.BaseType;
+            }
+            return builder.ToArray();
+        }
+
+        private static bool IsSuppressedByAttribute(PlatformAttributeInfo diagnosingAttribute, ISymbol containingSymbol)
+        {
+            while (containingSymbol != null)
+            {
+                var attributes = containingSymbol.GetAttributes();
+                if (attributes != null)
+                {
+                    foreach (AttributeData attribute in attributes)
+                    {
+                        if (diagnosingAttribute.AttributeType.ToString() == attribute.AttributeClass.Name)
+                        {
+                            if (PlatformAttributeInfo.TryParsePlatformAttributeInfo(attribute, out PlatformAttributeInfo parsedAttribute))
+                            {
+                                if (IsOSPlatformsEqual(diagnosingAttribute.PlatformName, parsedAttribute.PlatformName) &&
+                                    AttributeVersionsMatch(diagnosingAttribute, parsedAttribute))
+                                {
+                                    return true;
+                                }
+                            }
+                        }
+                    }
+                }
+                containingSymbol = containingSymbol.ContainingSymbol;
+            }
+            return false;
+        }
+
+        private static bool AttributeVersionsMatch(PlatformAttributeInfo diagnosingAttribute, PlatformAttributeInfo osAttribute)
+        {
+            return AttributeVersionsMatch(diagnosingAttribute.AttributeType, diagnosingAttribute.Version, osAttribute.Version);
+        }
+
+        private static bool AttributeVersionsMatch(PlatformAttributeType attributeType, Version diagnosingVersion, Version suppressingVersion)
+        {
+            if (attributeType == PlatformAttributeType.MinimumOSPlatformAttribute)
+            {
+                if (diagnosingVersion.Major != suppressingVersion.Major)
```

Comment:
Can we directly compare versions instead of individual parts of it?

Reply:
could be, even though I will keep it to avoid extra null check etc

Reply:
I meant can't we just use the version comparision operators: https://docs.microsoft.com/en-us/dotnet/api/system.version.op_lessthanorequal?view=netcore-3.1 and https://docs.microsoft.com/en-us/dotnet/api/system.version.op_greaterthanorequal?view=netcore-3.1. Why do we need to compare each part of the version?

Reply:
Because there was inconsistency with our logic and `Version` type comparison, but i have updated our Version creation logic, so now we can compare it directly, updated all